neuralop/training/offload.py

```diff
@@ -0,0 +1,228 @@
+    
+            * If tuple, specifies the output-shape of the **last** FNO Block
+    
+            * If tuple list, specifies the exact output-shape of each FNO Block
+        """
+        with torch.autograd.graph.save_on_cpu(pin_memory=True):
+            if output_shape is None:
+                output_shape = [None]*self.n_layers
+            elif isinstance(output_shape, tuple):
+                output_shape = [None]*(self.n_layers - 1) + [output_shape]
+    
+            # append spatial pos embedding if set
+            if self.positional_embedding is not None:
+                x = self.positional_embedding(x)
+            
+            x = self.lifting(x)
+    
+            if self.domain_padding is not None:
+                x = self.domain_padding.pad(x)
+    
+            for layer_idx in range(self.n_layers):
+                x = self.fno_blocks(x, layer_idx, output_shape=output_shape[layer_idx])
+        
+            if self.domain_padding is not None:
+                x = self.domain_padding.unpad(x)
+    
+            x = self.projection(x)
+    
+            return x
+        
+    FNO.forward = types.MethodType(offload_forward, FNO)
+    enable_activation_offload_for_FNOBlocks(FNO.fno_blocks)
+
+def enable_activation_offload_for_FNOBlocks(FNOBlocks):
+    
+    def forward_with_postactivation(self, x, index=0, output_shape=None):
+        with torch.autograd.graph.save_on_cpu(pin_memory=True):
+            x_skip_fno = self.fno_skips[index](x)
+            x_skip_fno = self.convs[index].transform(x_skip_fno, output_shape=output_shape)
+    
+            if self.use_channel_mlp:  
+                x_skip_channel_mlp = self.channel_mlp_skips[index](x)
+                x_skip_channel_mlp = self.convs[index].transform(x_skip_channel_mlp, output_shape=output_shape)
+    
+            if self.stabilizer == "tanh":
+                if self.complex_data:
+                    x = ctanh(x)
+                else:
+                    x = torch.tanh(x)
+    
+            x_fno = self.convs[index](x, output_shape=output_shape)
+            #self.convs(x, index, output_shape=output_shape)
+    
+            if self.norm is not None:
+                x_fno = self.norm[self.n_norms * index](x_fno)
+    
+            x = x_fno + x_skip_fno
+    
+            if (index < (self.n_layers - 1)):
+                x = self.non_linearity(x)
+    
+            if self.use_channel_mlp:  
+                x = self.channel_mlp[index](x) + x_skip_channel_mlp
+    
+            if self.norm is not None:
+                x = self.norm[self.n_norms * index + 1](x)
+    
+            if index < (self.n_layers - 1):
+                x = self.non_linearity(x)
+    
+            return x
+
+    FNOBlocks.forward_with_postactivation = types.MethodType(forward_with_postactivation, FNOBlocks)
+    for conv in FNOBlocks.convs:
+        enable_activation_offload_for_SpectralConv(conv)
+
+
+def enable_activation_offload_for_SpectralConv(SpectralConv):
```

Comment:
I wonder if this might be better as a module in `neuralop.layers`, or as a param to the original `SpectralConv`